### DISCUSSION TITLE
Phase 41.1 — AdversarialAttackSimulator: FGSM, PGD, C&W, AutoAttack Deep-Dive

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#830] [Show and tell]
# Phase 41.1 — AdversarialAttackSimulator: Implementation Deep-Dive

## FGSM: The Foundation

The Fast Gradient Sign Method (Goodfellow et al., 2015) remains the starting point for any adversarial robustness evaluation. Its elegance lies in its simplicity — a single gradient step in the direction that maximizes loss:

```
x_adv = x + ε · sign(∇_x L(f(x), y))
```

Our implementation extends this with:
- **Targeted FGSM**: Minimize loss toward a target class instead of maximizing true-class loss
- **Randomized FGSM**: Add uniform random initialization before the gradient step (Tramèr et al., 2018)
- **Momentum FGSM (MI-FGSM)**: Accumulate gradient momentum across iterations for better transferability

## PGD: The Gold Standard

Projected Gradient Descent (Madry et al., 2018) iteratively applies FGSM within an ε-ball:

```python
for step in range(num_steps):
    x_adv = x_adv + α · sign(∇_x L(f(x_adv), y))
    x_adv = project(x_adv, x, ε)  # project back to Lp ball
```

Key design decisions:
- **Random restarts** (default 10): Multiple random initializations to avoid local minima
- **Step size scheduling**: We support constant, decaying, and cyclic step sizes
- **Loss variants**: CE loss, DLR loss (Croce & Hein, 2020), and margin loss

## C&W Attack: Optimization-Based

The Carlini & Wagner (2017) L2 attack formulates adversarial example generation as an optimization problem:

```
minimize ‖δ‖₂ + c · f(x + δ)
```

Where f is a carefully chosen objective that is 0 when the attack succeeds. Key implementation details:
- **Tanh parameterization**: Ensures box constraints without projection
- **Binary search on c**: Finds minimal perturbation via binary search on the regularization constant
- **Adam optimizer**: Faster convergence than L-BFGS (original paper used L-BFGS)

## AutoAttack: Reliable Evaluation

AutoAttack (Croce & Hein, 2020) is a parameter-free ensemble of complementary attacks:

1. **APGD-CE**: Auto-PGD with cross-entropy loss (white-box)
2. **APGD-DLR**: Auto-PGD with Difference of Logits Ratio loss (white-box)
3. **FAB**: Fast Adaptive Boundary attack (white-box)
4. **Square**: Score-based black-box attack

The key innovation is the **automatic step size** in APGD — it adapts the step size based on the objective's progress, eliminating the need for manual tuning.

## Transferability Analysis

Our implementation includes systematic transferability analysis:
- Generate adversarial examples on a **source model**
- Evaluate success rate on **target model(s)**
- Track per-class transferability rates
- Support ensemble-based transfer (attacking multiple models simultaneously)

## Performance Benchmarks

| Attack | CIFAR-10 (ResNet-18) | Time/sample |
|--------|---------------------|-------------|
| FGSM (ε=8/255) | 47% robust acc | <1ms |
| PGD-20 (ε=8/255) | 0% robust acc | ~50ms |
| C&W L2 (untargeted) | 0% robust acc | ~2s |
| AutoAttack (ε=8/255) | 0% robust acc | ~5s |

*Results on naturally trained (non-robust) model — demonstrating the need for adversarial training (Phase 41.3).*

---
*See issue #825 for full specification. Part of [Phase 41 — Adversarial Robustness & Security Intelligence](../discussions/824).*
